### PR TITLE
add support for mongo 3 java driver

### PR DESCRIPTION
**Pull request by yairogen**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/foundation-runtime/cluster/pull/23" target="_blank">ReviewNinja

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java [hunk 1]
```diff
@@ -46,7 +45,7 @@ public void init(String id, String jobName) {
         this.jobName = jobName;
         this.masterSlaveLeaseTime = MasterSlaveConfigurationUtil.getMasterSlaveLeaseTime(jobName);
         if (isReady()) {
-            document = masterSlaveCollection.findOne(QueryBuilder.where(ID).equals(this.id));
+            document = (Document) masterSlaveCollection.find(new Document(ID,this.id)).limit(1).first();
```
Comment (igreenfield) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: Why not to use the queryBuilder?

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java [hunk 2]
```diff
@@ -61,13 +60,13 @@ public boolean isReady() {
     @Override
     public boolean isActiveVersion(String currentVersion) {
 
-        document = masterSlaveCollection.findOne(QueryBuilder.where(ID).equals(this.id));
+        document = (Document) masterSlaveCollection.find(new Document(ID,this.id)).limit(1).first();
```
Comment (igreenfield) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: What if there is no one?

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java [hunk 3]
```diff
@@ -81,58 +80,73 @@ public boolean isActiveVersion(String currentVersion) {
...
-        ConditionBuilder instanceIdEqaulity = QueryBuilder.where(MASTER_INSTANCE_ID).equals(MasterSlaveConfigurationUtil.INSTANCE_ID);
-        ConditionBuilder leaseCondition = QueryBuilder.where(LEASE_RENEWED).greaterThan(lastExpectedLeaseUpdateTime);
 
-        Document first = QueryBuilder.and(instanceIdEqaulity, leaseCondition);
-        ConditionBuilder second = QueryBuilder.where(LEASE_RENEWED).lessThanOrEqualTo(lastExpectedLeaseUpdateTime);
+        Bson one = Filters.eq(MASTER_INSTANCE_ID, MasterSlaveConfigurationUtil.INSTANCE_ID);
+        Bson two = Filters.eq(LEASE_RENEWED, lastExpectedLeaseUpdateTime);
+        Bson first = Filters.and(one, two);
+
+//        QueryBuilder first = QueryBuilder.start(MASTER_INSTANCE_ID)
```
Comment (igreenfield) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: Why all these comments?

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		MongoClientConfiguration config = new MongoClientConfiguration();
+		List<Pair<String, Integer>> mongodbServers = MasterSlaveConfigurationUtil.getMongodbServers();
+		for (Pair<String, Integer> mongodbServer : mongodbServers) {
+			config.addServer(mongodbServer.getLeft() + ":" + mongodbServer.getRight());
+		}
+    	config.setMaxConnectionCount(10);
+
+    	String dbName = MasterSlaveConfigurationUtil.getMongodbName();
+    	
+		if (MasterSlaveConfigurationUtil.isMongoAuthenticationEnabled()) {
```
Comment (igreenfield) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java]: What if the password is hashed?

**Review by igreenfield — APPROVED** [review]

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java (hunk 4 above)
Comment (yairogen) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java]: then user will provide a ready made client using the public connect(MongoClient mongoClient) API

Comment (igreenfield) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java]: it should be documented

Comment (yairogen) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/mongo/AsyncMongoClient.java]: ok

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java (hunk 3 above)
Comment (yairogen) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: removed.

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java (hunk 2 above)
Comment (yairogen) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: I create an empty one

File: master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java (hunk 1 above)
Comment (yairogen) [inline master-slave/src/main/java/com/cisco/oss/foundation/cluster/masterslave/mongo/MongoMastershipElector.java]: does it matter?

**Pull request merged**